### PR TITLE
core, partners: add token usage attribute to AIMessage

### PR DESCRIPTION
**Pull request by ccurme**
```python
class UsageMetadata(TypedDict):
    """Usage metadata for a message, such as token counts.

    Attributes:
        input_tokens: (int) count of input (or prompt) tokens
        output_tokens: (int) count of output (or completion) tokens
        total_tokens: (int) total token count
    """

    input_tokens: int
    output_tokens: int
    total_tokens: int
```
```python
class AIMessage(BaseMessage):
    ...
    usage_metadata: Optional[UsageMetadata] = None
    """If provided, token usage information associated with the message."""
    ...
```

### CODE REVIEW

File: libs/core/langchain_core/messages/ai.py
Comment (baskaryan) [inline libs/core/langchain_core/messages/ai.py:1]: will need to udpate chunk add logic as well

**Comment by ccurme** [comment]
Sample trace: https://smith.langchain.com/public/df479e96-412e-446a-a9fd-418a84a2d726/r

Screenshot 2024-05-21 at 11 06 22 AM

File: libs/core/langchain_core/messages/ai.py [hunk 1]
```diff
@@ -31,6 +47,11 @@ class AIMessage(BaseMessage):
     """If provided, tool calls associated with the message."""
     invalid_tool_calls: List[InvalidToolCall] = []
     """If provided, tool calls with parsing errors associated with the message."""
+    token_usage: Optional[TokenUsage] = None
```
Comment (dqbd) [inline libs/core/langchain_core/messages/ai.py:50]: Is there a reason why we§re not adding those fields in ChatGeneration instead?

**Comment by liugddx** [comment]
👀

**Comment by rishabh1024** [comment]
When can the the value of input_tokens be set to negative?

**Pull request merged**